plugins/catalog-backend/src/service/createRouter.ts

```diff
@@ -135,24 +137,139 @@ export async function createRouter(
-          filter: parseEntityFilterParams(req.query),
-          fields: parseEntityTransformParams(req.query),
-          order: parseEntityOrderParams(req.query),
-          pagination: parseEntityPaginationParams(req.query),
-          credentials: await httpAuth.credentials(req),
-        });
+        const filter = parseEntityFilterParams(req.query);
+        const fields = parseEntityTransformParams(req.query);
+        const order = parseEntityOrderParams(req.query);
+        const pagination = parseEntityPaginationParams(req.query);
+        const credentials = await httpAuth.credentials(req);
+
+        // When pagination parameters are passed in, use the legacy slow path
+        // that loads all entities into memory
+
+        if (pagination) {
+          const { entities, pageInfo } = await entitiesCatalog.entities({
+            filter,
+            fields,
+            order,
+            pagination,
+            credentials,
+          });
 
-        // Add a Link header to the next page
-        if (pageInfo.hasNextPage) {
-          const url = new URL(`http://ignored${req.url}`);
-          url.searchParams.delete('offset');
-          url.searchParams.set('after', pageInfo.endCursor);
-          res.setHeader('link', `<${url.pathname}${url.search}>; rel="next"`);
+          // Add a Link header to the next page
+          if (pageInfo.hasNextPage) {
+            const url = new URL(`http://ignored${req.url}`);
+            url.searchParams.delete('offset');
+            url.searchParams.set('after', pageInfo.endCursor);
+            res.setHeader('link', `<${url.pathname}${url.search}>; rel="next"`);
+          }
+
+          res.json(entities);
+          return;
         }
 
-        // TODO(freben): encode the pageInfo in the response
-        res.json(entities);
+        // For other read-the-entire-world cases, use queryEntities and stream
+        // out results.
+
+        // Imitate the httpRouter behavior of pretty-printing in development
+        const prettyPrint = process.env.NODE_ENV === 'development';
+
+        // The write lock is used for back pressure, preventing slow readers
+        // from forcing our read loop to pile up response data in userspace
+        // buffers faster than the kernel buffer is emptied.
+        // https://nodejs.org/api/http.html#http_response_write_chunk_encoding_callback
+        let writeLock: DeferredPromise | undefined;
+        const controller = new AbortController();
+        req.on('end', () => {
+          controller.abort();
+          writeLock?.resolve();
+          writeLock = undefined;
+        });
+        res.on('drain', () => {
+          writeLock?.resolve();
+          writeLock = undefined;
+        });
+
+        res.setHeader('Content-Type', 'application/json; charset=utf-8');
+        res.status(200);
+        res.flushHeaders();
+        res.write(prettyPrint ? '[\n' : '[');
+
+        const limit = 10000;
```

Comment:
Is there an opportunity to make this limit configurable?

Reply:
I try to keep the number of options to a minimum, just because they risk causing maintenance burden and confusion :) The exact value here isn't super important; it tries to strike a balance between memory use and streaming performance. The only reason that this number might get problematic, is if individual entities are so huge that even 10k of them are hard to fit in memory. But at that point you probably have huge issues 🤔